File: src/OliveMorocco.Business/Services/DocumentTotals.cs
namespace OliveMorocco.Business.Services;

public static class DocumentTotals
{
    private static decimal RoundMoney(decimal value) =>
        Math.Round(value, 2, MidpointRounding.AwayFromZero);

    public static decimal PuTtc(decimal prixUnitaireHt, decimal tauxTva) =>
        RoundMoney(prixUnitaireHt * (1 + tauxTva / 100m));

    public static decimal LineTtc(decimal quantite, decimal prixUnitaireHt, decimal remise, decimal tauxTva) =>
        RoundMoney(Math.Max(0, RoundMoney(quantite * PuTtc(prixUnitaireHt, tauxTva)) - remise));

    public static decimal LineHt(decimal quantite, decimal prixUnitaireHt, decimal remise, decimal tauxTva)
    {
        var rate = 1 + tauxTva / 100m;
        return rate == 0 ? 0 : RoundMoney(LineTtc(quantite, prixUnitaireHt, remise, tauxTva) / rate);
    }

    public static (decimal Ht, decimal Tva, decimal Ttc) Compute(
        IEnumerable<(decimal Quantite, decimal PrixUnitaireHT, decimal Remise, decimal TauxTVA)> lignes,
        decimal remiseGlobale = 0)
    {
        decimal ht = 0, tva = 0;
        foreach (var (quantite, prixUnitaireHt, remise, tauxTva) in lignes)
        {
            var lineTtc = LineTtc(quantite, prixUnitaireHt, remise, tauxTva);
            var lineHt = LineHt(quantite, prixUnitaireHt, remise, tauxTva);
            ht += lineHt;
            tva += lineTtc - lineHt;
        }

        ht = Math.Max(0, ht - remiseGlobale);
        return (ht, tva, ht + tva);
    }
}

File: src/OliveMorocco.Web/wwwroot/js/document-totals.js
(() => {
    const roundMoney = (value) => Math.round(value * 100) / 100;

    const parseNum = (value) => {
        const n = Number.parseFloat(value);
        return Number.isFinite(n) ? n : 0;
    };

    const puTtc = (pu, tva) => roundMoney(pu * (1 + tva / 100));

    const lineTtc = (qte, pu, remise, tva) =>
        roundMoney(Math.max(0, roundMoney(qte * puTtc(pu, tva)) - remise));

    const lineHt = (qte, pu, remise, tva) => {
        const rate = 1 + tva / 100;
        return rate === 0 ? 0 : roundMoney(lineTtc(qte, pu, remise, tva) / rate);
    };

    const lineTotalsForRow = (row, qte, pu, remise, tva) => ({
        ht: lineHt(qte, pu, remise, tva),
        ttc: lineTtc(qte, pu, remise, tva),
    });

    const renderLineTtcCells = (puTtcCell, ttcCell, qte, pu, remise, tva, formatMoney) => {
        const unitTtc = puTtc(pu, tva);
        const lineTtcValue = lineTtc(qte, pu, remise, tva);
        const unitAfter = qte > 0 ? roundMoney(lineTtcValue / qte) : unitTtc;

        if (puTtcCell)
            puTtcCell.textContent = formatMoney(unitAfter);
        if (ttcCell)
            ttcCell.textContent = formatMoney(lineTtcValue);
    };

    window.Zaho = window.Zaho || {};
    window.Zaho.DocumentTotals = {
        roundMoney,
        puTtc,
        lineTtc,
        lineHt,
        lineTotalsForRow,
        renderLineTtcCells,
    };
})();
